### DISCUSSION TITLE
Desert Sync does not work when no internet conection available

### DISCUSSION THREAD

**Discussion by n-rodrig** [NightscoutFoundation/xDrip#3356] [Bug reports]
First of all I would like to thank you for this magnific software.
Sencondly I am submitting a bug report because Desert Sync is not working properly on my device.If I am not mistaken in the documentation Desert Sync is designed to allow communication between master and follower on a network without internet connection. As long as the devices are seen on a LAN network, the follower should receive information from the master. 

I have 2 device a BQ Aquaris pro runing as Master (android 8) and a Samsung Galaxy M13 Runing as follower (android 14). Both devices are configured using desert sync using my home LAN. Both device have a fixed IP asigned by DHCP.  Follower recive data as long as my LAN is connected to internet but when Internet conection is disabled even if LAN is working properly the follower device recive no data. 

Configuration of master and follower have been done exactly as described in https://xdrip.readthedocs.io/en/latest/use/sync/?h=desert#desert-sync 

Why is this happening?
Are there any way to debug what is going wrong in follower comunication?

Other question is why _Google Play services_ are mandatory for Xdrip sync?

I include master follower status screenshoot.
![Follower status](https://github.com/NightscoutFoundation/xDrip/assets/12884810/5c3728a1-5d9f-4aa7-af0b-d08d77a01a39)
![Followerfollow  status](https://github.com/NightscoutFoundation/xDrip/assets/12884810/ec61cd05-e6c5-446c-8706-544c4d951e51)
![Folower status on no Active Internet conection](https://github.com/NightscoutFoundation/xDrip/assets/12884810/9248e27e-0d9d-4ced-a84c-ecba075f354f)
![MAster estatus](https://github.com/NightscoutFoundation/xDrip/assets/12884810/10152f71-bfb1-4309-9bb0-16539867c6b1)
![master follow status](https://github.com/NightscoutFoundation/xDrip/assets/12884810/518b9418-6ba3-436e-b2bd-e49136963b7e)

Thansks!

**Comment by Navid200** [comment]
I don't believe that is the definition of Desert Sync.  I believe this is:

If you are somewhere with no access to the network (no WiFi and no cell service), you can stay connected to someone who is physically close to you using Desert Sync.  Think of being on a motorbike ride with your follower sitting behind you.

**Comment by Navid200** [comment]
xDrip sync uses Google messaging, which requires Google play service.

**Comment by n-rodrig** [comment]
hi!
Thanks for information about google play services.  On the other hand, LAN connection without internet is one of the scenarios described on Desert sync documentation:
_"Desert Sync allows you to share data between your xDrip+ master and followers when internet is not available. This can happen when traveling out of phone coverage or on a local intranet (no internet access)."_

I have tried both LAN configured as described previosly and "Master as Wi-Fi hotspot" as described in documentation. Both cases only works while internet connection is available. 

_"If you're without mobile or Wi-Fi data, put the master phone in Wi-Fi hotspot mode and connect the followers through tethering. Maximum distance between devices will be limited"_

Why is this happening?
Are there any way to debug what is going wrong in follower comunication?

thanks!

**Comment by PhilTU** [comment]
I may be being confused by the way the information is presented, but are both showing the same IP address ?

Another possibility is that the router is prevented inter-device traffic "Wireless isolation, also called "client isolation," "AP isolation" and "SSID isolation," prevents users on a Wi-Fi network from reaching other devices in the same Wi-Fi network (SSID)" - some routers have this and it can be disabled. That might explain how it works when the internet is available but not if only using WLAN.

**Comment by n-rodrig** [comment]
Thanks PhilTU

   These are two interesting points:

1. Both Showing the same IP: I don´t really know what _is the meaning of
![imagen](https://github.com/NightscoutFoundation/xDrip/assets/12884810/a75120df-9d0f-49b6-b3ab-abb47653b78b)
 Follower IP is 192.168.1.12 master IP is 192168.1.18.  May be Master is thinking that follower have 192.168.1.18. If that is the case ¿how can I make identify follower real IP? 
 
2. I inter-device traffic is allowed, devices see each other on the net. Thats a good point but however inter-device trafic does not explain why hot spot configuration does not work.

**Reply by PhilTU** [reply]
Perhaps the intention is to show the interface the follower is accessed via, bit like a router table, so you don't see the IP of the follower but the IP through which the data flows to it. IDK. Perhaps if you had the internet access (ie a working setup) you could compare. I'll see if I can test that.

 -- edit -- I lashed up two phones and the Master's IP address shows on both. The phones are on different subnets so it was fairly obvious.

**Comment by n-rodrig** [comment]
Hi!

  Thanks for support! These are follower status page. I include both when internet is available and it works and what happens whe internet is nota available. 
![Follower-Status](https://github.com/NightscoutFoundation/xDrip/assets/12884810/e3b3139d-4404-4d58-963b-a6580913b4ca)

cheers!

**Reply by Navid200** [reply]
How far from each other are the two phones when it doesn't work?

**Reply by n-rodrig** [reply]
It does not matter the distace, both phones can be togeter but desert sync fail as soon as no internet connection is available.

**Reply by Navid200** [reply]
Can I ask how you verified that?
It is possible that something is broken since it was added.
But, that's the whole point of dessert sync, to have sync capability in the middle of a dessert where there is no Internet.

**Comment by alfureu** [comment]
just stumbled upon this discussion, and it made me think whether some sort of pseudo ad-hoc network cannot be created via bluetooth between 2 android phones. I mean every phone can connect to up to 5 devices, and one of them can be a follower phone where xDrip can follow the other phone's xDrip results. 

I am currently struggling to set up a Hotspot Wifi with my phone, as newer devices randomise these wifi networks, and also Wifi eats battery more than BLE Bluetooth

**Reply by n-rodrig** [reply]
Thanks, looks like an interesting question. HotSpot wifi consumes quite a bit and a bluethoot is already active on the main phone so for those who don't care about distance it's a better approach.
In my case I would like to have both devices connected no matter where I am in the house and, for that, a home wifi and desert sync seems to be the option. However I still haven´t made it works.

**Reply by alfureu** [reply]
desert sync work on my home wifi. However, I travel a lot, and enabling hotspot wifi is not ideal, as I move between locations, and I also have a smartwatch, so at some point when I join another wifi on one of the phones, the syncing between the 2 phones (master-follower) and smartwatch is broken. 

My request would be to have the option for desert sync not only via wifi, but possibly bluetooth as well

**Reply by Navid200** [reply]
I am currently traveling and have no access to a large screen.
When I get home in a few days, I can find out more.  But, if you like, you can have a look here:  

https://github.com/NightscoutFoundation/xDrip/issues/558

**Comment by alfureu** [comment]
Just a follow-up on this topic. I came across the project of [bluetooth_piconet](https://github.com/ilteoood/bluetooth_piconet) that might be of interest for desertsync via bluetooth. There are apparently other ways ([scatternet](https://en.wikipedia.org/wiki/Scatternet)) how to create a lan via bluetooth, but this is beyond my expertise, maybe the devs might look into this?

**Reply by Navid200** [reply]
Would there be an advantage using Bluetooth instead of WiFi hotspot?
For example, would you expect more range or less battery use?

**Reply by alfureu** [reply]
Yes, for example: one walks out of the home wifi setup (eg. travelling), and the follower phone can still sync with the master phone when no Wifi hotspot is available/not possible to set up. It would also use less energy (BLE vs Wifi), and make the moving outside of the local wifi and switching to bluetooth while travelling seamless (no need to set things up, ie. hotspot, every time)

**Reply by Navid200** [reply]
The point of desert sync is to be able to follow when neither the master nor the follower has access to the internet.  That is referred to as an internet desert.  
The requirement is for the two phones to be close so that the follower can be connected to the master.

We have tested this with WiFi hotspot.  But, if you establish connectivity between the two phones with Bluetooth, I don't see why it cannot be used as well.

This explains how to set up desert sync:
https://navid200.github.io/xDrip/docs/Follow/DesertSync.html

If you follow those instructions and experience a problem, please let us know.

**Reply by alfureu** [reply]
I am afraid the Bluetooth connection does not work this way by default (just tried it), as the 2 phones are not getting an IP. The whole request is to not to need an AP/Hotspot but just connect the phones via bluetooth, hence the suggestion to piconet/scatternet implementation, where such IP can be obtained via bluetooth not only Wifi

**Reply by Navid200** [reply]
> The whole request is to not to need an AP/Hotspot but just connect the phones via bluetooth 

That may be your request.  
If you look at the title of this discussion, you can see it obviously claims that you cannot use desert sync without access to the internet.  That is a false claim and I have tested it and it works, and I don't want anyone else finding this thread think the same.

You can open a new thread and make that request.  I don't know if it can be implemented.  But, that would be the first step.

**Reply by alfureu** [reply]
OK, gotcha, sorry for causing a confusion, the word "desert" in the DesertSync misled me

**Comment by PhilTU** [comment]
> I am afraid the Bluetooth connection does not work this way by default (just tried it), as the 2 phones..... where such IP can be obtained via bluetooth not only Wifi

Does Bluetooth tethering help? 
![Screenshot_20240321-173919](https://github.com/NightscoutFoundation/xDrip/assets/76240903/4058c2ea-218b-4a23-a57b-e1976456dab6)

**Reply by alfureu** [reply]
thanks for the heads-up, I can confirm this works. I wonder whether this works because there is a local network via bluetooth tethering, or because there is an access of the phone to mobile network and the data is being fed back from Nightscout. Case#1 would be in my reading DesertSync, Case#2 would be a nice coincidence but not addressing the DesertSync request. Can anyone confirm which is the case?

**Reply by Navid200** [reply]
Desert sync is there to allow the use of xDrip sync when there is no access to the web.
It is not meant to do the same with Nightscout.  There is no way Desert sync can give you access to a Nightscout server unless you are running your Nightscout server on your phone.

Of course if your master phone has access to the internet, connecting your follower to it, with WiFi hotspot or Bluetooth or whatever, will give it (follower) also access to the internet.  But, this has nothing to do with Desert sync.

**Reply by alfureu** [reply]
Hm, interesting. I actually have it differently: my master phone has no access to the internet, but bluetooth tethering connection to the follower phone, which in turn has access to the net (wifi) and hotspots this internet connection via bluetooth to the master 😎

**Reply by Navid200** [reply]
Actually, to correct myself, you won't even be using xDrip sync because xDrip sync needs access to Google servers.
Desert sync creates a local sync capability between a master and follower that are physically close enough so that the two phones can remain connected.

**Comment by Navid200** [comment]
> Hm, interesting. I actually have it differently: my master phone has no access to the internet, but bluetooth tethering connection to the follower phone, which in turn has access to the net (wifi) and hotspots this internet connection via bluetooth to the master 😎

If either phone has access to the web, establishing connectivity between the two phones using WiFi hotspot or Bluetooth will allow both phones to access the internet.
In that case, you can use any follower scheme.  You will even be able to use Dex share.

Desert sync can do something different.  It allows you to follow your partner/child/... when neither of you has access to the internet.

**Reply by alfureu** [reply]
Gotcha, I would just extend your sentence:

> Desert sync can do something different. It allows you to follow your partner/child/... when neither of you has access to the internet, **but one of the phones need to have hotspot enabled**

**Reply by Navid200** [reply]
Of course.  If neither phone has access to the internet, the only way one can follow the other is by having some form of connectivity.

If you read the description given by the person who created Desert Sync years ago, he has explained the whole thought behind it and how to enable it:
![Screenshot 2024-03-22 120346](https://github.com/NightscoutFoundation/xDrip/assets/51497406/2db21f72-cdc2-40e1-a2a1-fdec88e2ef5b)

![Screenshot 2024-03-22 120234](https://github.com/NightscoutFoundation/xDrip/assets/51497406/27912f80-1a55-4be1-a43e-df07f9cd03cc)
